File: dotnet/monosecret_dotnet/src/Monosecret/Monosecret.cs
namespace Monosecret;

/// <summary>Entry point for the Monosecret C# SDK.</summary>
public static class Monosecret
{
    /// <summary>Starts a fluent resolution builder.</summary>
    public static MonosecretBuilder Builder() => new();

    /// <summary>Resolves secrets in one call.</summary>
    public static Resolved Resolve(
        string? path = null,
        string? provider = null,
        string? profile = null,
        string? reason = null,
        string? scope = null,
        CallerContext? caller = null) =>
        Configured(path, provider, profile, scope, reason, caller).Load();

    /// <summary>Builds a value-free inventory report in one call.</summary>
    public static ResolutionReport Report(
        string? path = null,
        string? provider = null,
        string? profile = null,
        string? reason = null,
        string? scope = null,
        CallerContext? caller = null) =>
        Configured(path, provider, profile, scope, reason, caller).Report();

    /// <summary>The ABI version reported by the loaded native resolver.</summary>
    public static string AbiVersion() => Native.AbiVersion();

    private static MonosecretBuilder Configured(
        string? path,
        string? provider,
        string? profile,
        string? scope,
        string? reason,
        CallerContext? caller) =>
        Builder()
            .WithPath(path)
            .WithProvider(provider)
            .WithProfile(profile)
            .WithScope(scope)
            .WithReason(reason)
            .WithCaller(caller);
}

File: php/monosecret_php/src/Monosecret.php
<?php

declare(strict_types=1);

namespace Monosecret;

/**
 * Entry point for the Monosecret PHP SDK, mirroring the Rust derive crate's
 * `Monosecret::builder()`.
 *
 * The SDK is a thin client over the `monosecret_ffi` C ABI (loaded via PHP's
 * FFI extension): resolution — providers, fallback chains, profiles, generation,
 * `as_path` materialization — happens entirely in the Rust core, so every
 * provider works with no PHP-side logic.
 *
 * ```php
 * use Monosecret\Monosecret;
 *
 * $resolved = Monosecret::builder()
 *     ->withProvider('keyring://')
 *     ->withProfile('production')
 *     ->withReason('boot web app')
 *     ->load();
 *
 * echo $resolved->secrets['DATABASE_URL']->get();
 * $resolved->setAsEnv();
 * ```
 */
final class Monosecret
{
    /** Start a fluent {@see Builder}. */
    public static function builder(): Builder
    {
        return new Builder();
    }

    /**
     * Convenience one-shot resolve. Equivalent to building and calling
     * {@see Builder::load()}.
     *
     * @throws MissingRequiredException if a required secret is missing
     * @throws MonosecretException      for any other failure
     */
    public static function resolve(
        ?string $path = null,
        ?string $provider = null,
        ?string $profile = null,
        ?string $reason = null,
        ?string $scope = null,
        ?CallerContext $caller = null,
    ): Resolved {
        return self::configured($path, $provider, $profile, $scope, $reason, $caller)->load();
    }

    /**
     * Convenience one-shot value-free {@see Report}. Equivalent to building and
     * calling {@see Builder::report()}.
     *
     * @throws MonosecretException for a transport failure
     */
    public static function report(
        ?string $path = null,
        ?string $provider = null,
        ?string $profile = null,
        ?string $reason = null,
        ?string $scope = null,
        ?CallerContext $caller = null,
    ): Report {
        return self::configured($path, $provider, $profile, $scope, $reason, $caller)->report();
    }

    /** Build a {@see Builder} from the shared one-shot options. */
    private static function configured(
        ?string $path,
        ?string $provider,
        ?string $profile,
        ?string $scope,
        ?string $reason,
        ?CallerContext $caller,
    ): Builder {
        return self::builder()
            ->withPath($path)
            ->withProvider($provider)
            ->withProfile($profile)
            ->withScope($scope)
            ->withReason($reason)
            ->withCaller($caller);
    }

    /** The ABI version reported by the loaded native library. */
    public static function abiVersion(): string
    {
        return Native::abiVersion();
    }
}
